packages/Meshfree/src/DTK_DetailsSVDImpl.hpp

```diff
@@ -0,0 +1,304 @@
+/****************************************************************************
+ * Copyright (c) 2012-2017 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
+
+#ifndef DTK_DETAILS_SVD_IMPL_HPP
+#define DTK_DETAILS_SVD_IMPL_HPP
+
+#include <Kokkos_Core.hpp>
+
+#include <Kokkos_ArithTraits.hpp>
+
+namespace DataTransferKit
+{
+namespace Details
+{
+
+template <typename DeviceType>
+struct SVDFunctor
+{
+  public:
+    using ExecutionSpace = typename DeviceType::execution_space;
+
+    using matrices_type = Kokkos::View<double *, DeviceType>;
+    using shared_matrix =
+        Kokkos::View<double **, typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryUnmanaged>;
+    using ATS = Kokkos::ArithTraits<double>;
+    using matrix_2x2_type = Kokkos::Array<Kokkos::Array<double, 2>, 2>;
+
+  public:
+    SVDFunctor( int n, typename matrices_type::const_type As,
```

Comment:
`pseudoAs` are indeed the pseudo-inverses of `As`. It would be fine to use offsets in the future. I'm not sure if I like this, though. I would probably prefer to first group the matrices of the same size in blocks, and then work on blocks in parallel without offsets.

Reply:
@dalg24 Each matrix is $nxn$, but they all are flattened in an array. Thus, one can access $i$-th matrix as `[i*n*n, (i+1)*n*n)`.